File: LeetCode/src/dp/WordBreak.java
package dp;

import java.util.List;

public class WordBreak {
    public boolean wordBreak(String s, List<String> wordDict) {
        int n = s.length();
        boolean[] dp = new boolean[n + 1];
        dp[0] = true; // Empty string is always breakable

        for (int i = 1; i <= n; i++) {
            for (String word : wordDict) {
                int wordLen = word.length();
                if (i >= wordLen && dp[i - wordLen] && s.substring(i - wordLen, i).equals(word)) {
                    dp[i] = true;
                    break;
                }
            }
        }

        return dp[n];
    }

    public static void main(String[] args) {
        WordBreak solution = new WordBreak();
        String s = "leetcode";
        List<String> wordDict = List.of("leet", "code");
        System.out.println(solution.wordBreak(s, wordDict)); // Output: true
    }
}

File: LeetCode_Python/dp/strings/WordBreak.py
class WordBreak:
    def wordBreak(self, s: str, wordDict: list[str]) -> bool:
        n = len(s)
        dp = [False] * (n + 1)
        dp[0] = True

        for i in range(1, n + 1):
            for word in wordDict:
                length = len(word)
                if i >= length and s[i - length : i] == word and dp[i - length]:
                    dp[i] = True
                    break

        return dp[n]


# Time Complexity: O(n^2)
# Space Complexity: O(n)

if __name__ == "__main__":
    solution = WordBreak()

    print("Excepted Output: True")
    print("Actual Output: ", solution.wordBreak("leetcode", ["leet", "code"]))

    print("Excepted Output: True")
    print("Actual Output: ", solution.wordBreak("applepenapple", ["apple", "pen"]))

    print("Excepted Output: False")
    print(
        "Actual Output: ",
        solution.wordBreak("catsandog", ["cats", "dog", "sand", "and", "cat"]),
    )
